src/draw/lv_img_decoder.c

```diff
@@ -403,18 +447,75 @@ lv_res_t lv_img_decoder_built_in_open(lv_img_decoder_t * decoder, lv_img_decoder
 void lv_img_decoder_built_in_close(lv_img_decoder_t * decoder, lv_img_decoder_dsc_t * dsc)
 {
     LV_UNUSED(decoder); /*Unused*/
-    lv_img_dsc_t * img_dsc = (lv_img_dsc_t *)dsc->src;
-    lv_color_format_t cf = img_dsc->header.cf;
+    lv_img_decoder_built_in_data_t * decoder_data = dsc->user_data;
+    if(decoder_data) {
+        lv_free(decoder_data->img_data);
+        lv_free(decoder_data->palette);
+        lv_free(decoder_data);
+    }
+}
+
+lv_res_t lv_img_decoder_built_in_get_area(lv_img_decoder_t * decoder, lv_img_decoder_dsc_t * dsc,
+                                          const lv_area_t * full_area, lv_area_t * decoded_area)
+{
+    LV_UNUSED(decoder); /*Unused*/
+
+    lv_res_t res = LV_RES_INV;
+    lv_color_format_t cf = dsc->header.cf;
+    lv_img_decoder_built_in_data_t * decoder_data = dsc->user_data;
+    lv_coord_t w_px = lv_area_get_width(full_area);
+
     if(LV_COLOR_FORMAT_IS_INDEXED(cf)) {
-        lv_free((void *)dsc->img_data);
+        if(decoded_area->y1 == LV_COORD_MIN) {
+            dsc->img_data = lv_malloc(sizeof(lv_color32_t) * w_px);
+            LV_ASSERT_NULL(dsc->img_data);
+            if(dsc->img_data == NULL)
+                return LV_RES_INV;
+
+            *decoded_area = *full_area;
+            decoded_area->y2 = decoded_area->y1;
+        }
+        else {
+            decoded_area->y1++;
+            decoded_area->y2++;
+        }
+
+        if(decoded_area->y1 > full_area->y2) {
+            return LV_RES_INV;
+        }
+        else {
+            lv_coord_t bpp = lv_color_format_get_bpp(cf);
+            lv_coord_t x_fraction = decoded_area->x1 % (8 / bpp);
```

Comment:
`8 / bpp`
if bpp is greater than 8, it will cause a runtime problem.

Reply:
The guard of `if(LV_COLOR_FORMAT_IS_INDEXED(cf)) {` makes sure bpp is valid